Python ERSPublisher

StefanoVergani opened the pull request.

Commit:
first version of erspublisher and its test for Python

Commit:
first version of python protobuf stream test

Commit:
removed python protobuf

dingp committed:
syncing dunedaq-develop-cpp-ci.yml; previous commit: Merge: 154667c 166420c
Author: Marco Roda <31312964+mroda88@users.noreply.github.com>
Date:   Wed Jul 26 15:21:25 2023 +0100

    Merge pull request #19 from DUNE-DAQ/mroda/test_ers_with_causes

    Adding last issue with causes

dingp committed:
syncing ; previous commit: Author: Pengfei Ding <9438483+dingp@users.noreply.github.com>
Date:   Tue Aug 29 16:28:12 2023 -0500

    syncing dunedaq-develop-cpp-ci.yml; previous commit: Merge: 154667c 166420c
    Author: Marco Roda <31312964+mroda88@users.noreply.github.com>
    Date:   Wed Jul 26 15:21:25 2023 +0100

        Merge pull request #19 from DUNE-DAQ/mroda/test_ers_with_causes

        Adding last issue with causes

Commit:
minor changes

Commit:
newer version of the test, still not working but ERSPublisher is fixed

Commit:
new version

mroda88 committed:
Add better error handling

Commit:
perfectly working publisher test

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,57 @@
+        # Initialize Kafka producer
+        self.producer = KafkaProducer(
+            bootstrap_servers=self.bootstrap,
+            value_serializer=lambda v: v.SerializeToString(),
+            key_serializer=lambda k: str(k).encode('utf-8')
```
Potentially not necessary, especially considering that the value serialiser seems to be called twice in this case. Understand exactly what is happening is necessary.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,57 @@
+            key_serializer=lambda k: str(k).encode('utf-8')
+        )
+
+    def publish(self, issue, topic="ers_stream"):
```
Topic must be un internal configuration, it can be derived from the input configuration, but it cannot be exposed to users, the whole logic of this layer between user code and kafka is exactly to avoid confusion.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,57 @@
+        binary = issue.SerializeToString()
+        return self.producer.send(topic, key=issue.session, value=binary)
+
+    def close(self, timeout=None):
+        """Close the Kafka producer.
+
+        Args:
+            timeout (float, optional): The time to block during the close operation.
+        """
+        self.producer.close(timeout=timeout)
```
Potentially this should go, we don't want to raise kafka-level API to the user.

mroda88 commented on python/erskafka/__pycache__/ERSPublisher.cpython-310.pyc:
This file needs to go.

mroda88 commented on test/scripts/ers_python_publisher_test:
```diff
@@ -0,0 +1,70 @@
+import sys
+
+# I want it to point to sourcecode and not install
+desired_path = "/nfs/home/svergani/dev_area_2023_08_04/sourcecode/erskafka/python"
+sys.path.insert(0, desired_path)
```
No, get rid of these hardcoded.

mroda88 reviewed (changes requested):
Some things to be changed. We also need to think of ways to make this usage more friendly. Most likely we need an API that should allow us to generate the content of the context as automatically as possible. What info can be derived is to be investigated.

dingp committed:
syncing ; previous commit: Merge: 38e77e2 c7b6802
Author: Marco Roda <31312964+mroda88@users.noreply.github.com>
Date:   Thu Aug 31 14:03:24 2023 +0100

    Merge pull request #22 from DUNE-DAQ/mroda/improvements

    Add better error handling

plasorak commented:
We should talk about the packaging of this code. It should be installable with `pip` such that it can be installed in a microservice image. It's a relatively simple change here given the code isn't very complicated. I think this PR can go ahead and the code can be tracked here. We will need a new repository.

mroda88 committed:
Better error handling

mroda88 committed:
Add tag

Commit:
new changes after team's suggestions

Commit:
test new version

Commit:
newest version

Commit:
newest version

Commit:
new way of automatically creating errors in ERSPublisher

mroda88 commented on test/scripts/ers_python_publisher_test:
```diff
@@ -0,0 +1,27 @@
+        result = publisher.publish_simple_message(message)
+        print(f"Publishing issue {i}: {'successful' if result else 'failed'}")
+        sleep(0.5)
+
```
We need to test the exception as well

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+        """Destructor-like method to clean up resources."""
+        self.close()
+
+class ERSException(Exception):
```
NO, this is is not what I meant. We'll discuss it later.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+            package_name="erskafka"
+        )
+
+    def create_issue(self, message, name="DefaultIssueName", severity=None):
```
no, the default cannot be that. I don't think we need a default value of the name, if we have one, we need something more effective.

mroda88 replied:
More importantly, we need to add a possible cause.

mroda88 replied:
severity default to be "INFO", also, let's consider adding these as enum or the python equivalent.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+    def _generate_context(self):
+        """Private method to generate the context for an issue."""
+        return ersissue.Context(
+            cwd=os.getcwd(),
+            file_name=__file__,
+            function_name=inspect.currentframe().f_back.f_code.co_name,  # getting the caller function name
+            host_name=socket.gethostname(),
+            line_number=inspect.currentframe().f_back.f_lineno,  # getting the caller's line number
```
This is great!

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+            function_name=inspect.currentframe().f_back.f_code.co_name,  # getting the caller function name
+            host_name=socket.gethostname(),
+            line_number=inspect.currentframe().f_back.f_lineno,  # getting the caller's line number
+            package_name="erskafka"
```
We need to think about this.

mroda88 replied:
Two possible solutions
 - either we find a way from the code to find the name of the git repository of the line that calls the creation of the issue (which seems difficult to me)
 - Or we simply put something like "python-environment" or "unknown", in this case the decision depends on the other unknow info in the context.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+            host_name=socket.gethostname(),
+            line_number=inspect.currentframe().f_back.f_lineno,  # getting the caller's line number
+            package_name="erskafka"
+        )
```
What are the missing parameters? I think we should initialise them to something that makes clear the particular origins of these entries.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+        """Publish an ERS issue to the Kafka topic."""
+        return self.producer.send(self.topic, key=issue.session, value=issue)
+#this can be called by the user (it is best practise), but it is dealt by del if not
+    def close(self):
```
Regardless of the best practice, this will put the class in a unknown state, better to remove it.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,79 @@
+        issue = self.create_issue(message)
+        issue_chain = ersissue.IssueChain(
+            final=issue,
+            session=str(issue.time)  # using time as a session identifier
```
take it from enviromental variable: os.getenv('DUNEDAQ_PARTITION') but use "Unknown" as a default value.

mroda88 reviewed (changes requested):
Some logical changes to fix.

Commit:
context fixes

Commit:
new stuff

Commit:
new exceptions

Commit:
publisher test updated

Commit:
publisher test updated

mroda88 committed:
Draft

mroda88 committed:
More efficient way

mroda88 committed:
update tag

Commit:
ERSPublisher exceptions updated

Commit:
publisher test updated

Commit:
erspublisher new topic names

Commit:
erspublisher test new topic names

Commit:
erspublisher test new topic names

Commit:
erspublisher small bugs fixed

Commit:
some changes

Commit:
topic issue fixed

Commit:
timestamp fixed, code working perfectly

Commit:
function name now shows one step higher

Commit:
create_issue test added

Commit:
we know go up the source of issue

Commit:
now module is pointing to the source of the error

Commit:
exceptions fixed

Commit:
fixed repetition of execptions in output

Commit:
test with exceptions

Commit:
working

Commit:
working version

mroda88 committed:
sync with develop

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,158 @@
+    if exc:
+        # If the issue is created from an exception, set the name and inheritance
+        name = type(exc).__name__  # Use the exception's type name
+        inheritance_list = ["python_issue", "python_issue_from_exception", name]
```
The names of the hierarchy are classes, so let's use the class naming convention, and call them camel case: "PythonIssue", "PythonException", etc.

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,158 @@
+        issue = exception_to_issue(exc)  # Use the existing function to create an issue from the exception
+    else:
+        # For non-exception issues, continue as normal
+        inheritance_list = ["python_issue", name]
```
Same naming convention here

Commit:
camel case and warning info fixed

Commit:
camel case and warning info fixed

Commit:
redundancy in execption removed

Commit:
redundancy in execption removed

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,158 @@
+    )
+
+
+def create_issue(message, name="GenericPythonIssue", severity=SeverityLevel.INFO.value, exc=None):
```
The interface we want is not just an exception, it has to be a `cause`. The cause can be an `Exception` or another `SimpleIssue` or `IssueChain`

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,158 @@
+    return ersissue.SimpleIssue(
+        context=context,
+        message=str(exc),
+        time=current_time,
```
what unit is this? we need this to be in nanosecond. 
We'll discuss it tomorrow.

Commit:
time is now in nanoseconds

Commit:
time is now in nanoseconds

Commit:
now we have causes and not exc

Commit:
now we have causes and not exc

Commit:
cause appended to ihneritance removed, better use of cause

Commit:
changed details

Commit:
removed bugs

Commit:
newest version with issue.chain

mroda88 commented on python/erskafka/ERSPublisher.py:
```diff
@@ -0,0 +1,148 @@
+        elif isinstance(cause, ersissue.IssueChain):
+            # Set the final cause of the existing chain as the first cause of the new chain
+            issue_chain.causes.extend(cause.causes)
+            issue_chain.causes.append(cause.final)
```
cause.final is the first entry of the array.

mroda88 replied:
Followed up by cause.casuses without changing the order, as it is now.

mroda88 commented on test/scripts/ers_python_publisher_test:
```diff
@@ -0,0 +1,56 @@
+
+def cli(kafka_address, kafka_port, number_of_issues):
+    bootstrap = f"{kafka_address}:{kafka_port}"
+    conf = {"bootstrap": bootstrap, "topic": "monitoring.ers_stream"}
```
This topic is a bit misleading. As it is written, this will create a topic that is 
`monitoring.monitoring.ers_stream`. Which is not necessarily wrong, but it will be ignored by the microservice that puts the issue in the database. 

I understand the need for a non-disturbing topic for tests, I suggest this. 
 - Add an option with the topic name.
 - Default the option to `ers_stream` so that the default goes inside the database

mroda88 reviewed (changes requested):
I think this is all, we just need to fix these small things and test and see if anything along the line breaks.

Commit:
new causes

Commit:
topic name fixed

Commit:
inheritance reintroduced but it is not showing

Commit:
a non working ihneritance

Commit:
inheritance better but still not perfect

Commit:
inheritance finally working correctly

StefanoVergani committed:
Merge pull request #2 from DUNE-DAQ/develop

Sync PR

Commit:
wrong cli good idea

Commit:
Merge remote-tracking branch 'origin/svergani/erspublisher' into merge_branch

mroda88 committed:
Merge pull request #3 from StefanoVergani/merge_branch

Merge branch

Commit:
session almost working

Commit:
Merge remote-tracking branch 'origin/svergani/erspublisher' into svergani/erspublisher

Commit:
Revert "session almost working"

This reverts commit c8abbd7a9f2f15f7fec180619dd0f2e3966df182.

Commit:
session variable works perfectly

Commit:
latest version

Commit:
small issue in ihneritance fixed

Commit:
cleaning of the code

Commit:
further cleaning of comments

Commit:
further cleaning

Commit:
critical is now fatal in severity

Commit:
severity levels changed

Commit:
change in message

Commit:
changes in severity

Commit:
better comments

Commit:
better comments

Commit:
changed time management

mroda88 reviewed (approved):
Good first iteration
